### ISSUE TITLE
2.0.3: FUSE read error on graceful volume shutdown

### ISSUE THREAD

**Issue by onlyjob** [chrislusf/seaweedfs#1512]
A simple test setup: one master (started with `-defaultReplication=001`), one filer, two volume servers. Filer started with
 `-defaultReplicaPlacement=001 -encryptVolumeData -disableHttp`. Mounted file system with FUSE, copied 30 GB to the store, observed symmetric volume data growth on both volume servers. Run CRC check on data and during read operations stopped one volume server gracefully (with Ctrl+C). After 10 seconds, the very moment volume server stopped, got an `Read failed: Input/output error`.

 FUSE mount reported the following error:

~~~
fuse: panic in handler for Read [ID=0x17b830 Node=0x22 Uid=1000 Gid=1000 Pid=96109] 0x1 131072 @0x1607c000 dir=false fl=0 lock=0 ffl=OpenReadOnly: runtime error: slice bounds out of range [:114688] with capacity 0
~~~

During CRC check, both volume servers were active (I/O, CPU).
I'm not sure whether this problem is incomplete replication or lack of fail-over (to alternative/available volume) in FUSE client.
It does not matter which volume server I stop as stopping either of them result in read error on FUSE mount.

I tried to re-start CRC check with just one volume server but that failed with read error as well, despite explicit prior `volume.balance` and `volume.fix.replication` given in `weed shell` (both commands did not report anything).

It looks like data landed to both volume servers, that have identical size of "data" directory roughly matching the amount of data copied to FUSE mount. So replication appears to be working yet I could not read the whole data from FUSE mount when one volume server not running.

**Comment by chrislusf** [comment]
Thanks for the detailed report! 

Added a fix to read from alternative locations.

**Comment by onlyjob** [comment]
Thank you for responding so fast.  With a8624c2 situation is better but I was still able to reproduce read errors with one volume server down. Interestingly, some read operations occasionally succeed after retry.

**Comment by onlyjob** [comment]
Thank you for quick fix. With a8624c2 I could not reproduce the problem any more. All good, thanks.

**Comment by onlyjob** [comment]
Commit a8624c2 improved situation but after longer testing I've managed to reproduce the problem again...

**Comment by chrislusf** [comment]
> FUSE mount reported the following error:
> 
> ```
> fuse: panic in handler for Read [ID=0x17b830 Node=0x22 Uid=1000 Gid=1000 Pid=96109] 0x1 131072 @0x1607c000 dir=false fl=0 lock=0 ffl=OpenReadOnly: runtime error: slice bounds out of range [:114688] with capacity 0
> ```

I would need full logs from the mount. You can start with `weed -v=4 mount ...`

**Comment by onlyjob** [comment]
Thanks.

~~~~
I1009 13:32:56 19104 filehandle.go:209] doFlush /ISO/debian-installer/debian_stable_amd64/initrd.gz fh 15817943406903189082
I1009 13:32:56 19104 filehandle.go:177] Release /ISO/debian-installer/debian_stable_amd64/initrd.gz fh 15817943406903189082
I1009 13:32:56 19104 filehandle.go:209] doFlush /ISO/debian-installer/debian_stable_amd64/initrd.gz fh 15817943406903189082
I1009 13:32:56 19104 wfs.go:142] /ISO/debian-installer/debian_stable_amd64/initrd.gz ReleaseHandle id 15817943406903189082 current handles length 1
I1009 13:32:56 19104 dir.go:66] dir Attr /ISO/tecdoc03_04, attr: valid=1s ino=12942868949584229312 size=0 mode=drwxr-xr-x
I1009 13:32:56 19104 dir.go:233] dir Lookup /ISO/tecdoc03_04: TecDoc_03_04_6.iso by ID=0xb19d8 Node=0x11 Uid=1000 Gid=1000 Pid=25820
I1009 13:32:56 19104 dir.go:252] dir Lookup cache hit /ISO/tecdoc03_04/TecDoc_03_04_6.iso
I1009 13:32:56 19104 file.go:48] file Attr /ISO/tecdoc03_04/TecDoc_03_04_6.iso, open:0, existing attr: valid=1m0s ino=11393864575991920984 size=0 mode=-rw-r--r--
I1009 13:32:56 19104 file.go:91] file /ISO/tecdoc03_04/TecDoc_03_04_6.iso open Open [ID=0xb19da Node=0x19 Uid=1000 Gid=1000 Pid=25820] dir=false fl=OpenReadOnly
I1009 13:32:56 19104 wfs.go:116] AcquireHandle /ISO/tecdoc03_04/TecDoc_03_04_6.iso uid=1000 gid=1000
I1009 13:32:56 19104 file.go:97] /ISO/tecdoc03_04/TecDoc_03_04_6.iso file open handle id = 11393864575991920984
I1009 13:32:56 19104 filehandle.go:60] /ISO/tecdoc03_04/TecDoc_03_04_6.iso read fh 11393864575991920984: [0,16384) size 16384 resp.Data cap=16384
I1009 13:32:56 19104 reader_at.go:86] ReadAt [0,16384) of total file size 377417728 bytes 23 chunk views
I1009 13:32:56 19104 reader_at.go:118] read [0,16384), 0/23 chunk 3,076cdd40fe63 [0,15466496)
I1009 13:32:56 19104 reader_at.go:179] readFromWholeChunkData 3,076cdd40fe63 offset 0 [0,15466496) size at least 15466496
I1009 13:32:56 19104 reader_at.go:198] + doFetchFullChunkData 3,076cdd40fe63
I1009 13:32:56 19104 filechunk_manifest.go:99] read 3,076cdd40fe63 failed, err: fetch http://192.168.0.250:9334/3,076cdd40fe63: Get "http://192.168.0.250:9334/3,076cdd40fe63": dial tcp 192.168.0.250:9334: connect: connection refused
I1009 13:32:56 19104 reader_at.go:202] - doFetchFullChunkData 3,076cdd40fe63
2020/10/09 13:32:56 fuse: panic in handler for Read [ID=0xb19de Node=0x19 Uid=1000 Gid=1000 Pid=25820] 0x1 16384 @0x0 dir=false fl=0 lock=0 ffl=OpenReadOnly: runtime error: slice bounds out of range [:16384] with capacity 0
goroutine 369590 [running]:
github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).serve.func2(0x228e300, 0xc000129c20, 0xc000a4feb0, 0xc000a4fe26)
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:861 +0x1af
panic(0x1d7e3e0, 0xc001908680)
        runtime/panic.go:969 +0x175
github.com/chrislusf/seaweedfs/weed/filer.(*ChunkReadAt).doReadAt(0xc00019ec80, 0xc000c14000, 0x4000, 0x4000, 0x0, 0x4, 0x4, 0x92)
        github.com/chrislusf/seaweedfs/weed/filer/reader_at.go:125 +0xbee
github.com/chrislusf/seaweedfs/weed/filer.(*ChunkReadAt).ReadAt(0xc00019ec80, 0xc000c14000, 0x4000, 0x4000, 0x0, 0x0, 0x0, 0x0)
        github.com/chrislusf/seaweedfs/weed/filer/reader_at.go:87 +0x22e
github.com/chrislusf/seaweedfs/weed/filesys.(*FileHandle).readFromChunks(0xc0003c8180, 0xc000c14000, 0x4000, 0x4000, 0x0, 0x6, 0x6, 0x0)
        github.com/chrislusf/seaweedfs/weed/filesys/filehandle.go:123 +0x102
github.com/chrislusf/seaweedfs/weed/filesys.(*FileHandle).Read(0xc0003c8180, 0x22a2c40, 0xc000ed7900, 0xc000129c20, 0xc0012dcee0, 0x0, 0x0)
        github.com/chrislusf/seaweedfs/weed/filesys/filehandle.go:74 +0x2d7
github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).handleRequest(0xc0001c20c0, 0x22a2c40, 0xc000ed7900, 0x2267ea0, 0xc0001f8ba0, 0xc00164dc40, 0x228e300, 0xc000129c20, 0xc000a4feb0, 0xc00022e500, ...)
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:1255 +0xa75
github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).serve(0xc0001c20c0, 0x228e300, 0xc000129c20)
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:882 +0x2c7
github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).Serve.func1(0xc0001c20c0, 0x228e300, 0xc000129c20)
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:424 +0x6f
created by github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).Serve
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:422 +0x35f
I1009 13:32:56 19104 filehandle.go:60] /ISO/tecdoc03_04/TecDoc_03_04_6.iso read fh 11393864575991920984: [0,4096) size 4096 resp.Data cap=4096
I1009 13:32:56 19104 reader_at.go:86] ReadAt [0,4096) of total file size 377417728 bytes 23 chunk views
I1009 13:32:56 19104 reader_at.go:118] read [0,4096), 0/23 chunk 3,076cdd40fe63 [0,15466496)
I1009 13:32:56 19104 reader_at.go:179] readFromWholeChunkData 3,076cdd40fe63 offset 0 [0,15466496) size at least 15466496
I1009 13:32:56 19104 reader_at.go:198] + doFetchFullChunkData 3,076cdd40fe63
I1009 13:32:56 19104 filechunk_manifest.go:99] read 3,076cdd40fe63 failed, err: fetch http://192.168.0.250:9334/3,076cdd40fe63: Get "http://192.168.0.250:9334/3,076cdd40fe63": dial tcp 192.168.0.250:9334: connect: connection refused
I1009 13:32:56 19104 reader_at.go:202] - doFetchFullChunkData 3,076cdd40fe63
2020/10/09 13:32:56 fuse: panic in handler for Read [ID=0xb19e0 Node=0x19 Uid=1000 Gid=1000 Pid=25820] 0x1 4096 @0x0 dir=false fl=0 lock=0 ffl=OpenReadOnly: runtime error: slice bounds out of range [:4096] with capacity 0
goroutine 369591 [running]:
github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).serve.func2(0x228e300, 0xc000129d40, 0xc000eabeb0, 0xc000eabe26)
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:861 +0x1af
panic(0x1d7e3e0, 0xc000eb18e0)
        runtime/panic.go:969 +0x175
github.com/chrislusf/seaweedfs/weed/filer.(*ChunkReadAt).doReadAt(0xc00019ec80, 0xc00025d000, 0x1000, 0x1000, 0x0, 0x4, 0x4, 0x8f)
        github.com/chrislusf/seaweedfs/weed/filer/reader_at.go:125 +0xbee
github.com/chrislusf/seaweedfs/weed/filer.(*ChunkReadAt).ReadAt(0xc00019ec80, 0xc00025d000, 0x1000, 0x1000, 0x0, 0x0, 0x0, 0x0)
        github.com/chrislusf/seaweedfs/weed/filer/reader_at.go:87 +0x22e
github.com/chrislusf/seaweedfs/weed/filesys.(*FileHandle).readFromChunks(0xc0003c8180, 0xc00025d000, 0x1000, 0x1000, 0x0, 0x6, 0x6, 0x1ef545c)
        github.com/chrislusf/seaweedfs/weed/filesys/filehandle.go:123 +0x102
github.com/chrislusf/seaweedfs/weed/filesys.(*FileHandle).Read(0xc0003c8180, 0x22a2c40, 0xc000536c80, 0xc000129d40, 0xc001817680, 0x0, 0x0)
        github.com/chrislusf/seaweedfs/weed/filesys/filehandle.go:74 +0x2d7
github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).handleRequest(0xc0001c20c0, 0x22a2c40, 0xc000536c80, 0x2267ea0, 0xc0001f8ba0, 0xc00164dc40, 0x228e300, 0xc000129d40, 0xc000eabeb0, 0xc0005e3440, ...)
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:1255 +0xa75
github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).serve(0xc0001c20c0, 0x228e300, 0xc000129d40)
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:882 +0x2c7
github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).Serve.func1(0xc0001c20c0, 0x228e300, 0xc000129d40)
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:424 +0x6f
created by github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs.(*Server).Serve
        github.com/chrislusf/seaweedfs/vendor/github.com/seaweedfs/fuse/fs/serve.go:422 +0x35f
I1009 13:32:56 19104 filehandle.go:209] doFlush /ISO/tecdoc03_04/TecDoc_03_04_6.iso fh 11393864575991920984
I1009 13:32:56 19104 filehandle.go:177] Release /ISO/tecdoc03_04/TecDoc_03_04_6.iso fh 11393864575991920984
I1009 13:32:56 19104 filehandle.go:209] doFlush /ISO/tecdoc03_04/TecDoc_03_04_6.iso fh 11393864575991920984
I1009 13:32:56 19104 wfs.go:142] /ISO/tecdoc03_04/TecDoc_03_04_6.iso ReleaseHandle id 11393864575991920984 current handles length 1
I1009 13:32:56 19104 dir.go:233] dir Lookup /: .git by ID=0xb19e6 Node=0x1 Uid=1000 Gid=1000 Pid=26738
I1009 13:32:56 19104 dir.go:233] dir Lookup /: .git by ID=0xb19e8 Node=0x1 Uid=1000 Gid=1000 Pid=26738
I1009 13:32:56 19104 dir.go:233] dir Lookup /: HEAD by ID=0xb19ea Node=0x1 Uid=1000 Gid=1000 Pid=26738
~~~~

It looks like the the problem is due to `dial tcp 192.168.0.250:9334: connect: connection refused`.

In another attempt I've got similar error when another volume server was down: `filechunk_manifest.go:99] read 2,08e956d8e460 failed, err: fetch http://192.168.0.2:9335/2,08e956d8e460: Get "http://192.168.0.2:9335/2,08e956d8e460": dial tcp 192.168.0.2:9335: connect: connection refused`. 

Strangely enough after retrying 3...4 times read operation progressed further but failed a little bit later with a similar error.

---

Also I think SeaweedFS `mount` should be more resilient to failures. For example, when both volume servers are briefly unavailable (like in case of rebooting a network switch), a read operation should not fail instantly. Instead client should retry operation long enough, indeally for a configurable number of tries before failing. LizardFS and MooseFS implemented that concept with `mfsioretries` mount option: https://manpages.debian.org/buster/lizardfs-client/mfsmount.1.en.html

NFS keeps trying infinitely, until connectivity with server is restored.

**Comment by chrislusf** [comment]
Thanks! The logs are very helpful! Fixed one bug.

In https://github.com/chrislusf/seaweedfs/commit/8d34eb005087f6cb305b8934d8614a88b62cff99 I also added exponentially backoff if read error happens, for about 10 minutes. Currently the wait time is hard coded.

**Comment by chrislusf** [comment]
FYI: Previous 8d34eb0 had a dead loop bug. Fixed now.

**Comment by onlyjob** [comment]
> And reduced the total retry time to about 20 seconds.

IMHO this is a bad idea as retry time should be much longer by default...

Eventually, could you make it configurable with a mount option please?

The point of delay is to avoid recoverable errors. Some applications may not recover gracefully after read errors (e.g. MySQL). Distributed storage systems are designed for reliability. By default, admin should be able to reboot a volume server or two (e.g. in case of brief power failure) without risk of applications failing. Reboot might easily take longer than 20 minutes (e.g. server decides to `fsck` 20 local HDDs, etc).

On MooseFS/LizardFS retries are not time-timed but a number of retries can be configured on FUSE client. I set the number to a pretty high value (around 500) to buy enough time to fix a problem like reboot of network switch (due to firmware update, etc.). Even restarting a network switch might take half an hour or longer if you have to apply (or rollback) a firmware update.

I would set default time for retries in FUSE client to at least 30 minutes.
On my cluster I would change it to 6 hours or even longer.

Only in exceptional cases client should throw a read error - like when all replicas are permanently lost.
In such cases, those who want to return error quick (e.g. on dev/staging cluster) should be able to lower the default retry time but for most situations, where reliability matters, FUSE client should retry as long as possible, even indefinitely (like NFS), giving storage admin a chance to fix the problem.

Ideally FUSE client should throw a read error _only_ when replicas are really lost (e.g. when all known volume servers are up but data is missing). FUSE client should just wait while volume server is down until it becomes available again or until storage admin marks the volume server as lost.

**Comment by chrislusf** [comment]
added `weed mount -readRetryTime` https://github.com/chrislusf/seaweedfs/commit/d155f907c2759ce8c14f50952fd7ea4a68bcac65

**Comment by onlyjob** [comment]
Chris, you are a legend.  :)  Thank you.

**Comment by onlyjob** [comment]
Fixed in 2.0.4. While reading files from FUSE mount I've stopped one volume server, then stopped another one, then started the first volume server and read operations did not fail. Well done. Thanks.
